File: jpush/android/src/main/java/com/capacitorjs/plugins/jpush/JPushPlugin.java
package com.capacitorjs.plugins.jpush;

import android.content.Intent;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.PermissionCallback;
import com.getcapacitor.util.PermissionHelper;

import java.util.HashSet;
import java.util.Set;

import cn.jpush.android.api.JPushInterface;
import cn.jpush.android.api.JPushMessage;
import cn.jpush.android.api.NotificationMessage;
import cn.jpush.android.ups.JPushUPSManager;

@CapacitorPlugin(name = "JPush")
public class JPushPlugin extends Plugin {

    private JPush implementation = new JPush();

    public static JPushPlugin instance;
    public String initCallId;
    public String tagsCallId;
    public String aliasCallId;
    @Override
    public void load() {
        super.load();
        instance = this;
    }

    @Override
    protected void handleOnResume() {
        super.handleOnResume();
        if (this.initCallId != null) {
            PluginCall call = this.bridge.getSavedCall(this.initCallId);
            int status = JPushInterface.isNotificationEnabled(getContext());
            JSObject res = new JSObject();
            if (status == 1) {
                JPushUPSManager.registerToken(this.getContext(),  this.getConfig().getString("appKey"), null, null, result -> {
                    if (result.getReturnCode() == 0) {
                        res.put("data", "authorized");
                    } else {
                        res.put("data", "denied");
                    }
                    call.resolve(res);
                    this.bridge.releaseCall(call);
                    this.initCallId = null;
                });
            } else {
                res.put("data", "denied");
                call.resolve(res);
                this.bridge.releaseCall(call);
                this.initCallId = null;
            }
        }
    }

    @PluginMethod
    public void init(PluginCall call) {
        JPushInterface.setDebugMode(!this.getConfig().getBoolean("isProduction", true));
        JPushInterface.setChannel(this.getContext(), this.getConfig().getString("channel", "default"));
        int status = JPushInterface.isNotificationEnabled(getContext());
        JSObject res = new JSObject();
        if (status == -1) {
            res.put("data", "denied");
            call.resolve(res);
        } else if (status == 0) {
            JPushInterface.requestRequiredPermission(this.getActivity());
            this.initCallId = call.getCallbackId();
            this.bridge.saveCall(call);
        } else if (status == 1) {
            JPushUPSManager.registerToken(this.getContext(),  this.getConfig().getString("appKey"), null, null, result -> {
                if (result.getReturnCode() == 0) {
                    res.put("data", "authorized");
                } else {
                    res.put("data", "denied");
                }
                call.resolve(res);
            });
        }
    }

    @PluginMethod
    public void getRegistrationID(PluginCall call) {
        String registrationID = JPushInterface.getRegistrationID(this.getContext());
        JSObject res = new JSObject();
        JSObject data = new JSObject();
        data.put("registrationID", registrationID);
        res.put("data", data);
        call.resolve(res);
    }

    @PluginMethod
    public void setTags(PluginCall call) {
        JSArray tags = call.getArray("tags");
        Set<String> resultSet = new HashSet<>();
        for (int i = 0; i < tags.length(); i++) {
            String tag = tags.optString(i);
            resultSet.add(tag);
        }
        Integer sequence = call.getInt("sequence");
        JPushInterface.setTags(getContext(), sequence, resultSet);
        this.tagsCallId = call.getCallbackId();
        this.bridge.saveCall(call);
    }

    @PluginMethod
    public void setAlias(PluginCall call) {
        JPushInterface.setAlias(getContext(), call.getInt("sequence"), call.getString("alias"));
        this.aliasCallId = call.getCallbackId();
        this.bridge.saveCall(call);
    }

    static void onTagOperatorResult(JPushMessage jPushMessage) {
        if (instance.tagsCallId != null) {
            JSObject res = new JSObject();
            JSObject data = new JSObject();
            int errorCode =  jPushMessage.getErrorCode();
            data.put("iResCode", errorCode == 0 ? 1 : 0);
            data.put("iTags", new JSArray(jPushMessage.getTags()));
            data.put("seq", jPushMessage.getSequence());
            res.put("data", data);
            PluginCall call = instance.bridge.getSavedCall(instance.tagsCallId);
            call.resolve(res);
            instance.bridge.releaseCall(call);
            instance.tagsCallId = null;
        }
    }

    static void onAliasOperatorResult(JPushMessage jPushMessage) {
        if (instance.aliasCallId != null) {
            JSObject res = new JSObject();
            JSObject data = new JSObject();
            int errorCode =  jPushMessage.getErrorCode();
            data.put("iResCode", errorCode == 0 ? 1 : 0);
            data.put("iAlias", jPushMessage.getAlias());
            data.put("seq", jPushMessage.getSequence());
            res.put("data", data);
            PluginCall call = instance.bridge.getSavedCall(instance.aliasCallId);
            call.resolve(res);
            instance.bridge.releaseCall(call);
            instance.aliasCallId = null;
        }
    }

    static void onNotifyMessageOpened(NotificationMessage notificationMessage) {
        Intent intent = new Intent(instance.getContext(), instance.getActivity().getClass());
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        instance.getActivity().startActivity(intent);
        JSObject res = new JSObject();
        JSObject data = new JSObject();
        data.put("badge", 0);
        data.put("body", notificationMessage.inAppMsgContentBody);
        data.put("subtitle", notificationMessage.notificationContent);
        data.put("title", notificationMessage.notificationTitle);
        data.put("extras", notificationMessage.notificationExtras);
        res.put("data", data);
        instance.notifyListeners("openNotification", res);
    }
}

File: jpush/ios/Sources/JPushPlugin/JPushPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(JPushPlugin)
public class JPushPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "JPushPlugin"
    public let jsName = "JPush"
    var userinfo: [AnyHashable : Any]?
    var authCallbackId: String?

    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "init", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRegistrationID", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAlias", returnType: CAPPluginReturnPromise),
    ]
    private let implementation = JPush()

    override public func load() {   
        NotificationCenter.default.addObserver(self, selector: #selector(applicationDidLaunch), name: UIApplication.didFinishLaunchingNotification, object: nil)
    }
    @objc func `init`(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            #if targetEnvironment(simulator)
            call.resolve(["data": ""])
            #else
            self.authCallbackId = call.callbackId
            self.bridge?.saveCall(call)
            #endif
            self.startJPushSDK()
        }
    }
    @objc func getRegistrationID(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            JPUSHService.registrationIDCompletionHandler { resCode, registrationID in
                call.resolve(["data": ["code": resCode,
                                       "registrationID": registrationID ?? ""]])
            }
        }
    }
    @objc func setTags(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let tags = call.getArray("tags") as! [String]
            JPUSHService.setTags(Set(tags), completion: { iResCode, iTags, seq in
                let data = [
                    "iResCode": iResCode,
                    "iTags": iTags?.compactMap{$0 as? String} ?? [],
                    "seq": seq
                ]
                call.resolve(["data": data])
            }, seq: call.getInt("sequence") ?? 0)
        }
    }
    @objc func setAlias(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            JPUSHService.setAlias(call.getString("alias")!, completion: { iResCode, iAlias, seq in
                call.resolve(["data": [
                    "iResCode": iResCode,
                    "iAlias": iAlias! as String,
                    "seq": seq
                ]])
            }, seq: call.getInt("sequence") ?? 0)
        }
    }
    @objc func applicationDidLaunch (notification: Notification) {
        self.userinfo = notification.userInfo
        JPUSHService.setDebugMode()
    }
    @objc func startJPushSDK() {
        let entity = JPUSHRegisterEntity()
        entity.types = NSInteger(UNAuthorizationOptions.alert.rawValue) |
        NSInteger(UNAuthorizationOptions.sound.rawValue) |
        NSInteger(UNAuthorizationOptions.badge.rawValue)

        JPUSHService.register(forRemoteNotificationConfig: entity, delegate: self)
        JPUSHService.setup(withOption: self.userinfo, appKey: self.getConfig().getString("appKey") ?? "", channel: self.getConfig().getString("channel") ?? "", apsForProduction: self.getConfig().getBoolean("isProduction", false), advertisingIdentifier: nil)
    }
}

extension JPushPlugin: JPUSHRegisterDelegate {
    
    private func openNotification(request: UNNotificationRequest) {
        let extras = request.content.userInfo
        let content = request.content

        let badge = content.badge
        let body = content.body
        let subtitle = content.subtitle
        let title = content.title
        self.notifyListeners("openNotification", data: ["badge": badge ?? 0, "body": body, "subtitle": subtitle, "title": title, "extras": extras])
    }
    //MARK - JPUSHRegisterDelegate
    @available(iOS 10.0, *)
    public func jpushNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: (() -> Void)) {
        let userInfo = response.notification.request.content.userInfo
        if (response.notification.request.trigger?.isKind(of: UNPushNotificationTrigger.self) == true) {
            JPUSHService.handleRemoteNotification(userInfo)
        }
        self.openNotification(request: response.notification.request)
        completionHandler()
    }
    
    @available(iOS 10.0, *)
    public func jpushNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification,
                                 withCompletionHandler completionHandler: ((Int) -> Void)) {
        completionHandler(Int(UNNotificationPresentationOptions.badge.rawValue | UNNotificationPresentationOptions.sound.rawValue | UNNotificationPresentationOptions.alert.rawValue))
    }
    
    public func jpushNotificationCenter(_ center: UNUserNotificationCenter, openSettingsFor notification: UNNotification) {
    }
    public func jpushNotificationAuthorization(_ status: JPAuthorizationStatus, withInfo info: [AnyHashable : Any]?) {
        var data = "notDetermined"
        if status == .statusDenied {
            data = "denied"
        } else if status == .statusAuthorized {
            data = "authorized"
        } else if status == .statusProvisional {
            data = "provisional"
        }
        if let callId = self.authCallbackId {
            let call = self.bridge?.savedCall(withID: callId);
            call?.resolve(["data": data])
            self.bridge?.releaseCall(withID: callId)
            self.authCallbackId = nil;
        }
    }
}
